### PR TITLE
[RLMRealm] Make inWriteTransaction public

### LINKED ISSUE #1541: Make inWriteTransaction public
I have the following category on `RLMObject` in my app:

```objective-c
- (void)app_transactionWithBlock:(void(^)(RLMRealm *realm, id model))block
{
    NSAssert(block, @"Block is required.");
    RLMRealm *realm = [RLMRealm defaultRealm];
    if (!realm.inWriteTransaction) {
        [realm transactionWithBlock:^{
            block(realm, self);
        }];
    } else {
        block(realm, self);
    }
}
```

It would have been very handy if `inWriteTransaction` was public. Thanks!

### PR DESCRIPTION
**Pull request by segiddins**
Closes #1541.

\c @jpsim

### CODE REVIEW

File: RealmSwift/Tests/RealmTests.swift [hunk 1]
```diff
@@ -163,6 +163,19 @@ class RealmTests: TestCase {
         XCTAssertEqual(Realm().objects(SwiftStringObject).count, 0)
     }
 
+    func testInWriteTransaction() {
```
Comment (jpsim) [inline RealmSwift/Tests/RealmTests.swift]: Please test `Realm().invalidate()`

**Comment by jpsim** [comment]
Please add an Objective-C test.

I think the documentation should warn readers about the potential misuse or anti-patterns that this new property enables. That's the whole reason why we never made this public in the first place.

**Comment by segiddins** [comment]
> I think the documentation should warn readers about the potential misuse or anti-patterns that this new property enables. That's the whole reason why we never made this public in the first place.

What exactly would those be, though?

**Comment by jpsim** [comment]
Perhaps this:

```objc
/**
Indicates if this Realm is currently in a write transaction.

@warning Wrapping mutating operations in a write transaction if this property returns `NO`
         may cause a large number of write transactions to be created, which could negatively
         impact Realm's performance. Always prefer performing multiple mutations in a single
         transaction when possible.
*/
```

File: RealmSwift/Tests/RealmTests.swift (hunk 1 above)
Comment (segiddins) [inline RealmSwift/Tests/RealmTests.swift]: done

**Comment by jpsim** [comment]
:+1: CI failure is unrelated. I'll spend some time once Swift is released taming CI's flakiness.

**Comment by astigsen** [comment]
Are there use cases for this method that goes beyond just being a
workaround for the lack of nested transactions? If not, we might want to
consider if we want to add methods that we know will have to be deprecated
later when the real feature is added.

**Comment by tgoyne** [comment]
Mostly just bad ones, but there's clearly no chance of getting nested transactions for 1.0.

**Comment by segiddins** [comment]
Updated.

**Comment by jpsim** [comment]
:+1:

**Comment by tgoyne** [comment]
:+1:

**Comment by segiddins** [comment]
CI failure is totally unrelated.

**Pull request merged**